p2p/enr/enr.go

```diff
@@ -0,0 +1,293 @@
+// also implement rlp.Decoder if additional checks are needed on the value.
+type Key interface {
+	ENRKey() string
+}
+
+type pair struct {
+	k []byte
+	v []byte
+}
+
+type Record struct {
+	seq       uint32 // sequence number
+	signature []byte // record's signature
+	raw       []byte // RLP encoded record
+	pairs     []pair // list of all key/value pairs, sorted prior to RLP encoding
+	signed    bool   // keeps track if record was modified after it was signed and encoded
+}
+
+func (r Record) Seq() uint32 {
+	return r.seq
+}
+
+func (r *Record) SetSeq(s uint32) {
+	r.signed = false
+	r.seq = s
+}
+
+func (r *Record) Load(k Key) (bool, error) {
+	for _, p := range r.pairs {
+		if string(p.k) == k.ENRKey() {
+			err := rlp.DecodeBytes(p.v, k)
+			return true, err
+		}
+	}
+
+	return false, errors.New("record does not exist")
+}
+
+func (r *Record) Set(k Key) error {
+	r.signed = false
+	blob, err := rlp.EncodeToBytes(k)
+	if err != nil {
+		return err
+	}
+	r.pairs = append(r.pairs, pair{[]byte(k.ENRKey()), blob})
+	return nil
+}
+
+func (r Record) EncodeRLP(w io.Writer) error {
+	if !r.signed {
+		return errors.New("record is not signed")
+	}
+
+	_, err := w.Write(r.raw)
+
+	return err
+}
+
+func (r *Record) DecodeRLP(s *rlp.Stream) error {
+	var err error
+
+	r.signature, err = s.Bytes()
+	if err != nil {
+		return err
+	}
+
+	_, err = s.List()
+	if err != nil {
+		return err
+	}
+
+	if err := s.Decode(&r.seq); err != nil {
+		return err
+	}
+
+	// read key/value pairs until we reach rlp.EOL
+	for _, _, err = s.Kind(); err == nil; _, _, err = s.Kind() {
+		key, err2 := s.Bytes()
+		if err2 != nil {
+			return err2
+		}
+
+		value, err2 := s.Bytes()
+		if err2 != nil {
+			return err2
+		}
+
+		r.pairs = append(r.pairs, pair{k: key, v: value})
+	}
+
+	if err != rlp.EOL {
+		return err
+	}
+
+	sigcontent, err := r.serialisedContent()
+	if err != nil {
+		return err
+	}
+
+	// update r.raw
+	blob, err := rlp.EncodeToBytes(r.signature)
+	if err != nil {
+		return err
+	}
+
+	r.raw = append(blob, sigcontent...)
+
+	err = r.verifySignature(sigcontent)
+	if err != nil {
+		return err
+	}
+
+	// mark record ready for encoding
+	r.signed = true
+
+	return nil
+}
+
+func (r Record) Equal(o Record) (bool, error) {
+	rr, err := r.serialisedContent()
+	if err != nil {
+		return false, err
+	}
+
+	oo, err := o.serialisedContent()
+	if err != nil {
+		return false, err
+	}
+
+	if bytes.Compare(rr, oo) != 0 {
+		return false, nil
+	}
+
+	if err := r.verifySignature(rr); err != nil {
+		return false, err
+	}
+
+	if err := o.verifySignature(oo); err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (r *Record) NodeAddr() ([]byte, error) {
+	var secp256k1 Secp256k1
+
+	_, err := r.Load(&secp256k1)
+	if err != nil {
+		return nil, err
+	}
+
+	digest := crypto.Keccak256Hash(secp256k1)
+
+	return digest.Bytes(), nil
+}
+
+func (r *Record) Sign(privkey *ecdsa.PrivateKey) error {
+	r.seq = r.seq + 1
+
+	id := ID(ID_SECP256k1_KECCAK)
+
+	r.Set(id)
```

Comment:
You can be less verbose and write `r.Set(ID(ISECP256k1_KECCAK))`.

Reply:
Done.